### PR TITLE
815 - Ensure only one portal metadata computed file exists at a time

### PR DESCRIPTION
**Pull request by nozomione**
## Issue Number

Closing #815

Feature branch: `feature/portal-metadata-command`

## Purpose/Implementation Notes
The following methods were added to the `create_portal_metadata` management command:
- `purge_computed_file` for purging the portal metadata computed file
Adjusted  the`test_create_portal_metadata` test:
- Added new `test test_purge_computed_file`

## Types of changes
- New feature (non-breaking change which adds functionality)

## Functional tests

N/A

## Checklist

- [x] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Screenshots

N/A

### CODE REVIEW

File: api/scpca_portal/management/commands/create_portal_metadata.py [hunk 1]
```diff
@@ -20,11 +20,15 @@ def add_arguments(self, parser):
         parser.add_argument(
             "--clean-up-output-data", action=BooleanOptionalAction, default=settings.PRODUCTION
         )
+        parser.add_argument("--delete-from-s3", action=BooleanOptionalAction, default=False)
         parser.add_argument(
             "--update-s3", action=BooleanOptionalAction, default=settings.UPDATE_S3_DATA
         )
+        parser.add_argument("--purge", action=BooleanOptionalAction, default=False)
```
Comment (davidsmejia) [inline api/scpca_portal/management/commands/create_portal_metadata.py:27]:
```suggestion
```

File: api/scpca_portal/management/commands/create_portal_metadata.py [hunk 2]
```diff
@@ -20,11 +20,15 @@ def add_arguments(self, parser):
         parser.add_argument(
             "--clean-up-output-data", action=BooleanOptionalAction, default=settings.PRODUCTION
         )
+        parser.add_argument("--delete-from-s3", action=BooleanOptionalAction, default=False)
```
Comment (davidsmejia) [inline api/scpca_portal/management/commands/create_portal_metadata.py:23]:
```suggestion
```

File: api/scpca_portal/management/commands/create_portal_metadata.py [hunk 3]
```diff
@@ -46,3 +50,16 @@ def create_portal_metadata(self, **kwargs):
...
+
+    def purge_computed_file(self, delete_from_s3=False):
+        logger.info("Purging the portal-wide metadata computed file")
+        computed_file = ComputedFile.objects.filter(portal_metadata_only=True).first()
+
+        if computed_file:
+            if delete_from_s3:
+                logger.info("Deleting the zip from S3")
+                s3.delete_output_file(computed_file.s3_key)
+            logger.info("Deleting the instance from the database")
```
Comment (davidsmejia) [inline api/scpca_portal/management/commands/create_portal_metadata.py:62]:
```suggestion
            logger.info("Deleting object from the database")
```

File: api/scpca_portal/management/commands/create_portal_metadata.py [hunk 4]
```diff
@@ -46,3 +50,16 @@ def create_portal_metadata(self, **kwargs):
                 computed_file.clean_up_local_computed_file()
 
         return computed_file
+
+    def purge_computed_file(self, delete_from_s3=False):
+        logger.info("Purging the portal-wide metadata computed file")
+        computed_file = ComputedFile.objects.filter(portal_metadata_only=True).first()
+
+        if computed_file:
+            if delete_from_s3:
```
Comment (davidsmejia) [inline api/scpca_portal/management/commands/create_portal_metadata.py:59]:
```suggestion
    def purge_computed_file(self,  computed_file, delete_from_s3=False):
        logger.info("Purging the portal-wide metadata computed file")

        if update_s3_data:
```

File: api/scpca_portal/management/commands/create_portal_metadata.py [hunk 5]
```diff
@@ -20,11 +20,15 @@ def add_arguments(self, parser):
...
         )
+        parser.add_argument("--delete-from-s3", action=BooleanOptionalAction, default=False)
         parser.add_argument(
             "--update-s3", action=BooleanOptionalAction, default=settings.UPDATE_S3_DATA
         )
+        parser.add_argument("--purge", action=BooleanOptionalAction, default=False)
 
     def handle(self, *args, **kwargs):
+        if kwargs("purge", False):
+            self.purge_computed_file(delete_from_s3=kwargs.get("delete_from_s3", False))
```
Comment (davidsmejia) [inline api/scpca_portal/management/commands/create_portal_metadata.py:31]:
```suggestion
```

File: api/scpca_portal/management/commands/create_portal_metadata.py [hunk 6]
```diff
@@ -20,11 +20,15 @@ def add_arguments(self, parser):
...
         self.create_portal_metadata(**kwargs)
 
     def create_portal_metadata(self, **kwargs):
```
Comment (davidsmejia) [inline api/scpca_portal/management/commands/create_portal_metadata.py:34]: here you would add:

```python
if computed_file := ComputedFile.objects.filter(portal_metadata_only=True).first():
    self.purge_computed_file(computed_file, delete_from_s3)

```

File: api/scpca_portal/test/management/commands/test_create_portal_metadata.py [hunk 7]
```diff
@@ -111,3 +111,31 @@ def test_create_portal_metadata(self, mock_upload_output_file):
...
+
+            # Purge computed_file
+            self.processor.purge_computed_file(delete_from_s3=True)
+            mock_delete_output_file.assert_called_once_with(computed_file.s3_key)
+            # Make sure that computed_file has been deleted from the database
+            computed_file_after_purge = ComputedFile.objects.filter(
+                id=computed_file_before_purge.id, portal_metadata_only=True
+            ).first()
+            self.assertIsNone(computed_file_after_purge)
+        else:
```
Comment (davidsmejia) [inline api/scpca_portal/test/management/commands/test_create_portal_metadata.py:140]: `test_only_one_computed_file_at_any_point` could be the test name, and you just need to run the command twice, then ensure that there is only one record in the database for computed files that has `portal_metadata_only=True`

File: api/scpca_portal/management/commands/create_portal_metadata.py (hunk 4 above)
Comment (nozomione) [reply api/scpca_portal/management/commands/create_portal_metadata.py]: See at 119d43b0469b3157c846f2e73200467c6a154c32.

File: api/scpca_portal/test/management/commands/test_create_portal_metadata.py (hunk 7 above)
Comment (nozomione) [reply api/scpca_portal/test/management/commands/test_create_portal_metadata.py]: See at 9b19c58cad3b77feaafca2d726224d13a8c6d932.

File: api/scpca_portal/management/commands/create_portal_metadata.py (hunk 6 above)
Comment (nozomione) [reply api/scpca_portal/management/commands/create_portal_metadata.py]: See at 9b19c58cad3b77feaafca2d726224d13a8c6d932.

File: api/scpca_portal/management/commands/create_portal_metadata.py (hunk 5 above)
Comment (nozomione) [reply api/scpca_portal/management/commands/create_portal_metadata.py]: See at 9b19c58cad3b77feaafca2d726224d13a8c6d932.

File: api/scpca_portal/management/commands/create_portal_metadata.py (hunk 3 above)
Comment (nozomione) [reply api/scpca_portal/management/commands/create_portal_metadata.py]: See at 9b19c58cad3b77feaafca2d726224d13a8c6d932.

File: api/scpca_portal/management/commands/create_portal_metadata.py (hunk 2 above)
Comment (nozomione) [reply api/scpca_portal/management/commands/create_portal_metadata.py]: See at 9b19c58cad3b77feaafca2d726224d13a8c6d932.

File: api/scpca_portal/management/commands/create_portal_metadata.py (hunk 1 above)
Comment (nozomione) [reply api/scpca_portal/management/commands/create_portal_metadata.py]: See at 9b19c58cad3b77feaafca2d726224d13a8c6d932.

**Comment by nozomione** [comment]
I've applied your feedback and this PR is ready for your review. Thank you David!

**Comment by nozomione** [comment]
I've applied your input and this PR is ready for another look. Thank you David!

File: api/scpca_portal/management/commands/create_portal_metadata.py [hunk 8]
```diff
@@ -46,3 +50,11 @@ def create_portal_metadata(self, clean_up_output_data: bool, update_s3: bool, **
             computed_file.clean_up_local_computed_file()
 
         return computed_file
+
+    def purge_computed_file(self, computed_file, update_s3=False):
+        logger.info("Purging the portal-wide metadata computed file")
```
Comment (davidsmejia) [inline api/scpca_portal/management/commands/create_portal_metadata.py:55]:
```suggestion
```

**Review by davidsmejia — COMMENTED** [review]
Left a suggestion for more explicit checking of calling the command twice so we don't have to worry about the return value from the command method.

File: api/scpca_portal/test/management/commands/test_create_portal_metadata.py [hunk 9]
```diff
@@ -121,3 +121,20 @@ def test_create_portal_metadata(self, mock_upload_output_file):
...
+            clean_up_output_data=False, update_s3=True
+        )
+        # Second call to create the portal metadata computed file
+        self.processor.create_portal_metadata(clean_up_output_data=False, update_s3=True)
+        mock_delete_output_file.assert_called_with(
+            computed_file_one.s3_key, computed_file_one.s3_bucket
+        )
+        # Make sure pre-existing computed_file has been deleted and only one exists
+        existing_computed_file = ComputedFile.objects.filter(portal_metadata_only=True)
+        self.assertEqual(existing_computed_file.count(), 1)
```
Comment (davidsmejia) [inline api/scpca_portal/test/management/commands/test_create_portal_metadata.py:140]: 1. lets call the command twice without storing the return value
2. move the existing_computed_files count check above the mock_delete_ouput_file check
3. grab what should be the only computed file for portal metadata and assert that mock_delete_ouput_file was called with that object's values

**Review by davidsmejia — APPROVED** [review]
LGTM

**Pull request merged**